### PR TITLE
Quickcons via ajax

### PR DESCRIPTION
**Pull request by chmst**
### Summary of Changes
Quickicons read now data via an ajax request. 
The respective controllers in components have got a method for delivering data. 
Buttons can be selected via params.

### CODE REVIEW

**Comment by chmst** [comment]
Use getTotal() for icon amounts in users and privacy

**Comment by brianteeman** [comment]
>  Then the quickicon plugins can be removed

Removing the general import for the quickicon group means you're removing support for extensions to add their own entries to that list.  Akeeba or K2 or JSitemap Pro are just some of the extensions which add items to this list.

**Comment by brianteeman** [comment]
This really is nothing to do with the template and it should only be dealt with in the main repo where everyone can see what is happening and "problems" like this will be spotted earlier - saving everyone time

File: administrator/components/com_cache/Controller/DisplayController.php [hunk 1]
```diff
@@ -66,6 +68,36 @@ public function display($cachable = false, $urlparams = false)
...
+		if (!empty($data))
+		{
+			foreach ($data as $d)
+			{
+				$size += $d->size;
+			}
+		}
+
+		// Number bytes are returned in format xxx.xx MB
+		$mb = explode(' ', HTMLHelper::_('number.bytes', $size, 'MB', 1, false));
```
Comment (bembelimen) [inline administrator/components/com_cache/Controller/DisplayController.php:95]: ```$result = strstr(HTMLHelper::_('number.bytes', $size, 'MB', 1, false), '.', true);```

File: administrator/components/com_categories/Controller/CategoriesController.php [hunk 2]
```diff
@@ -39,6 +40,24 @@ public function getModel($name = 'Category', $prefix = 'Administrator', $config
...
+	 *
+	 * @since   4.0
+	 */
+	public function getQuickiconContent()
+	{
+		$model = $this->getModel('Categories');
+		$model->setState('filter.published', 1);
+		$model->setState('filter.extension', 'com_content');
+
+		$amount = $model->getTotal() ? $model->getTotal() : '0';
```
Comment (bembelimen) [inline administrator/components/com_categories/Controller/CategoriesController.php:56]: ```$amount = (int) $model->getTotal();```

File: administrator/components/com_checkin/Controller/DisplayController.php [hunk 3]
```diff
@@ -89,6 +89,8 @@ public function getMenuBadgeData()
 
 		$model = $this->getModel('Checkin');
 
-		echo new JsonResponse(count($model->getItems()));
+		$amount = count($model->getItems()) ? count($model->getItems()) : '0';
```
Comment (bembelimen) [inline administrator/components/com_checkin/Controller/DisplayController.php:92]: ```$amount = count($model->getItems());```

File: administrator/components/com_content/Controller/ArticlesController.php [hunk 4]
```diff
@@ -133,4 +134,22 @@ public function getModel($name = 'Article', $prefix = 'Administrator', $config =
...
+	 *
+	 * @since   4.0
+	 */
+	public function getQuickiconContent()
+	{
+		$model = $this->getModel('articles');
+		
+		$model->setState('filter.published', 1);
+
+		$amount = $model->getTotal() ? $model->getTotal() : '0';
```
Comment (bembelimen) [inline administrator/components/com_content/Controller/ArticlesController.php:151]: ```$amount = (int) $model->getTotal();```

File: administrator/components/com_menus/Controller/ItemsController.php [hunk 5]
```diff
@@ -59,6 +60,25 @@ public function getModel($name = 'Item', $prefix = 'Administrator', $config = ar
...
+	 * @since   4.0
+	 */
+	public function getQuickiconContent()
+	{
+		$model = $this->getModel('Items');
+
+		$model->setState('filter.published', 1);
+		$model->setState('filter.client_id', 0);
+
+		$amount = $model->getTotal() ? $model->getTotal() : '0';
```
Comment (bembelimen) [inline administrator/components/com_menus/Controller/ItemsController.php:77]: ```$amount = $model->getTotal();```

File: administrator/components/com_modules/Controller/ModulesController.php [hunk 6]
```diff
@@ -71,4 +72,23 @@ public function getModel($name = 'Module', $prefix = 'Administrator', $config =
...
+	 * @since   4.0
+	 */
+	public function getQuickiconContent()
+	{
+		$model = $this->getModel('Modules');
+
+		$model->setState('filter.published', 1);
+		$model->setState('filter.client_id', 0);
+
+		$amount = $model->getTotal() ? $model->getTotal() : '0';
```
Comment (bembelimen) [inline administrator/components/com_modules/Controller/ModulesController.php:90]: ```$amount = $model->getTotal();```

File: administrator/components/com_plugins/Controller/PluginsController.php [hunk 7]
```diff
@@ -35,4 +36,22 @@ public function getModel($name = 'Plugin', $prefix = 'Administrator', $config =
...
+	 *
+	 * @since   4.0
+	 */
+	public function getQuickiconContent()
+	{
+		$model = $this->getModel('Plugins');
+
+		$model->setState('filter.enabled', 1);
+
+		$amount = $model->getTotal() ? $model->getTotal() : '0';
```
Comment (bembelimen) [inline administrator/components/com_plugins/Controller/PluginsController.php:53]: ```$amount = $model->getTotal();```

File: administrator/components/com_users/Controller/UsersController.php [hunk 8]
```diff
@@ -152,4 +153,22 @@ public function activate()
...
+	 *
+	 * @since   4.0
+	 */
+	public function getQuickiconContent()
+	{
+		$model = $this->getModel('Users');
+
+		$model->setState('filter.state', 0);
+
+		$amount = $model->getTotal() ? $model->getTotal() : '0';
```
Comment (bembelimen) [inline administrator/components/com_users/Controller/UsersController.php:170]: ```$amount = $model->getTotal();```

File: administrator/modules/mod_quickicon/Helper/QuickIconHelper.php [hunk 9]
```diff
@@ -73,417 +63,184 @@ public static function &getButtons(Registry $params, CMSApplication $application
...
-				$context = $params->get('context', 'mod_quickicon');
-
-				// Include buttons defined by published quickicon plugins
-				PluginHelper::importPlugin('quickicon');
+				PluginHelper::importPlugin('quickicon', 'joomlaupdate');
+			}
+			if ($params->get('show_eupdate', '1'))
+			{
+				PluginHelper::importPlugin('quickicon', 'extensionupdate');
+			}
```
Comment (bembelimen) [inline administrator/modules/mod_quickicon/Helper/QuickIconHelper.php:73]: I would add a blank line for readability

File: administrator/modules/mod_quickicon/Helper/QuickIconHelper.php [hunk 10]
```diff
@@ -73,417 +63,184 @@ public static function &getButtons(Registry $params, CMSApplication $application
...
+			if ($params->get('show_oupdate', '1'))
+			{
+				PluginHelper::importPlugin('quickicon', 'overridecheck');
+			}
```
Comment (bembelimen) [inline administrator/modules/mod_quickicon/Helper/QuickIconHelper.php:77]: I would add a blank line for readability

File: administrator/modules/mod_quickicon/Helper/QuickIconHelper.php [hunk 11]
```diff
@@ -73,417 +63,184 @@ public static function &getButtons(Registry $params, CMSApplication $application
...
+						'image'   => null,
+						'text'    => null,
+						'name'    => null,
+						'linkadd' => null,
+						'access'  => true,
+						'class'   => true,
+						'group'   => 'MOD_QUICKICON',
+					);
+
+					$icon = array_merge($default, $icon);
```
Comment (bembelimen) [inline administrator/modules/mod_quickicon/Helper/QuickIconHelper.php:103]: I would add a blank line for readability

**Review by bembelimen — APPROVED** [review]

**Comment by chmst** [comment]
Thank you! Code improvements will follow in a new PR

**Pull request merged**